src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java

```diff
@@ -231,21 +233,37 @@ public Event getActiveCustodialEvent(Long offenderId) {
         return activeCustodyConvictions.get(0);
     }
 
-    @SuppressWarnings("unused")
     @Transactional
-    public Custody addOrReplaceOrDeleteCustodyKeyDates(Long offenderId, Long convictionId, ReplaceCustodyKeyDates replaceCustodyKeyDates) {
+    public Custody addOrReplaceOrDeleteCustodyKeyDates(@SuppressWarnings("unused") Long offenderId, Long convictionId, ReplaceCustodyKeyDates replaceCustodyKeyDates) {
         var event = eventRepository.findById(convictionId).orElseThrow();
 
-        //noinspection StatementWithEmptyBody
         if (updateCustodyKeyDatesFeatureSwitch) {
-            // TODO the actual update
+            final var custodyManagedKeyDates = custodyManagedKeyDates();
+            final var missingKeyDateTypesCodes = missingKeyDateTypesCodes(replaceCustodyKeyDates);
+            event
+                    .getDisposal()
+                    .getCustody()
+                    .getKeyDates()
+                    .stream()
+                    .map(KeyDate::getKeyDateType)
+                    .map(StandardReference::getCodeValue)
+                    .filter(custodyManagedKeyDates::contains)  // all key dates managed by this service
+                    .filter(missingKeyDateTypesCodes::contains) // all ones missing from request
+                    .collect(toList()) // collect into new list so we can start deleting
+                    .forEach(keyDate -> deleteCustodyKeyDate(event, keyDate));
```

Comment:
Possibly move this into a method to document it better?

Reply:
I might do in next iteration; I know I have to refactor this when I deal with creating a Contact note